engine/src/witness/sol/egress_witnessing.rs

```diff
@@ -0,0 +1,114 @@
+use crate::sol::{
+	commitment_config::CommitmentConfig,
+	retry_rpc::{SolRetryRpcApi, SolRetryRpcClient},
+	rpc_client_api::{
+		RpcTransactionConfig, TransactionConfirmationStatus, TransactionStatus,
+		UiTransactionEncoding,
+	},
+};
+use anyhow::Result;
+use cf_chains::sol::{SolSignature, LAMPORTS_PER_SIGNATURE};
+use itertools::Itertools;
+
+pub async fn get_finalized_fee(
+	sol_client: &SolRetryRpcClient,
+	signature: SolSignature,
+) -> Result<u64> {
+	match sol_client
+		.get_signature_statuses(&[signature], false)
+		.await
+		.value
+		.iter()
+		.exactly_one()
+		.expect("We queried for exactly one signature.")
+	{
+		Some(TransactionStatus {
+			confirmation_status: Some(TransactionConfirmationStatus::Finalized),
+			..
+		}) => {
+			let transaction_meta = sol_client
+				.get_transaction(
+					&signature,
+					RpcTransactionConfig {
+						encoding: Some(UiTransactionEncoding::Json),
+						// Using finalized there could be a race condition where this
+						// doesn't get the tx. But "Processed" is timing out so we better
+						// retry with finalized.
+						commitment: Some(CommitmentConfig::finalized()),
+						// Getting also type 0 even if we don't use them atm
+						max_supported_transaction_version: Some(0),
+					},
+				)
+				.await
+				.transaction
+				.meta;
+
+			Ok(match transaction_meta {
+				Some(meta) => meta.fee,
+				// This shouldn't happen. Want to avoid Erroring. We either default to
+				// 5000 or return OK(()) so we don't submit transaction_succeeded and
+				// retry again later. Defaulting to avoid potentially getting stuck not
+				// witness something because no meta is returned.
+				None => LAMPORTS_PER_SIGNATURE,
+			})
+		},
+		Some(TransactionStatus { confirmation_status: other_status, .. }) => Err(anyhow::anyhow!(
+			"Transaction status is {:?}, waiting for {:?}.",
+			other_status,
+			TransactionConfirmationStatus::Finalized
+		)),
+		None => Err(anyhow::anyhow!("Unknown Transaction.")),
```

Comment:
Wouldn't this be reasonably likely? This is what we discussed yesterday, that it'll log error for now but we can improve it later?

Reply:
Yes - will add a comment to clarify.